.changeset/wet-hoops-drop.md

```diff
@@ -0,0 +1,17 @@
+---
+"fluid-framework": minor
+"@fluidframework/tree": minor
+---
+
+nodeChanged event now includes the list of properties that changed
+
+The payload of the `nodeChanged` event emitted by SharedTree now includes a `changedProperties` property that indicates
+which properties of the node changed.
+
+For object nodes, the list of properties uses the property identifiers defined in the schema, and not the persisted
+identifiers (or "stored keys") that can be provided through `FieldProps` when defining a schema.
+See the documentation for `FieldProps` for more details about the distinction between "view keys" and "stored keys".
+
+For map nodes, every key that was added, removed, or updated by a change to the tree is included in the list of properties.
+
+For array nodes, the list of properties will always be empty.
```

Comment:
I'm concerned we may want to reverse this decision when we finally expose data about what the changes are.
Either that, or we'll want to offer a different event signature for arrays than we do for objects and maps.
Whichever we choose, it will be a walk back from the direction this PR is taking.
It's not the end of the world to deprecate one event and create a new one, but it's creates noise both for us and our users, so it'd be nice to avoid it if we can.

Reply:
If we wanted to set up something today that can be just expanded (instead of replaced) in the future, I think we'd have to go with a `Map<string, Record<string, unknown>>`? Today we can provide the keys but no other details, so empty values. Eventually replace them with payloads that have the appropriate properties depending on what we expose and how. I find a map with no values kind of weird as an API for today, but I'm not totally opposed.

Reply:
I think the fact that `changedProperties` is inside of a property bag `{...}` (as opposed to directly being the first parameter to the event) is sufficient for future-proofing and about as good of a guess as we can make. To add more information in the future, we don't need to break the event signature, we just add more properties to the event bag. Maybe there's some overlap with the existing properties and a new property supersedes an old one, but that's fine, the user can just ignore the deprecated one.

Reply:
I don't follow what we gain from advertising a map value of type `Record<string, unknown>`. Are you picturing putting an empty object in there for now? What if we want the map value to be something like `RangeChange[]` (`RangeChange` doesn't exist, I'm just making it up for illustration) instead, is that compatible with `Record<string, unknown>`?

Reply:
Empty object for now, yeah. I'd imagine that for future-proofing, we'd want the payload to be a property bag anyway like Noah pointed out, so the hypothetical `RangeChange[]` would be in a property of that bag.